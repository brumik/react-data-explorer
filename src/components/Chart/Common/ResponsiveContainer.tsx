import React, {
    FunctionComponent,
    useEffect,
    useRef,
    useState
} from 'react';
import {
    getApiData,
    getLegendData
} from '../Api';
import { ChartFetchFunction } from '../Functions/types';
import {
    ChartApiProps,
    ChartApiData
} from '../types';
import ErrorState from './ErrorState';

interface Props {
    setWidth: (width: number) => void
    height: number,
    api: ChartApiProps,
    setData: (data: ChartApiData) => void,
    fetchFnc: ChartFetchFunction
}

const ResponsiveContainer: FunctionComponent<Props> = ({
    setWidth,
    height,
    api,
    setData,
    children,
    fetchFnc
}) => {
    const containerRef = useRef<HTMLDivElement>(null);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState(false);

    const handleResize = () => {
        if (containerRef.current && containerRef.current.clientWidth) {
            setWidth(containerRef.current.clientWidth);
        }
    };

    useEffect(() => {
        handleResize();
        window.addEventListener('resize', handleResize);

        return () => {
            window.removeEventListener('resize', handleResize);
        }
    }, []);

    useEffect(() => {
        let didCancel = false;

        setLoading(true);
        getApiData(api, fetchFnc)
            .then(results => {
                if (!didCancel) {
                    setError(false);
                    setData({ ...results, legend: getLegendData(results)});
                }
            })
            .catch(() => {
                if (!didCancel) {
                    setError(true);
                }
            })
            .finally(() => {
                if (!didCancel) {
                    setLoading(false);
                }
            });

        return () => {
            didCancel = true;
        };
    }, [api])

    return (
        <div ref={containerRef}>
            <div style={{ height }}>
                { !loading && !error && children }
                { !loading && error && <ErrorState />}
            </div>
        </div>
    );
}

export default ResponsiveContainer;