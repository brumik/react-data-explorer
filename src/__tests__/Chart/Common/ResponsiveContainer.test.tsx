import React from 'react'
import { render, act, waitFor } from '@testing-library/react';
import fetchMock from 'fetch-mock-jest';
import ResponsiveContainer from '../../../components/Chart/Common/ResponsiveContainer';
import fetchFnc from '../../../components/Chart/Functions/fetchFnc';
import { pieChartResponse } from '../../assets/fetchResponses';

describe('Chart/Common/ResponsiveContainer', () => {
    afterEach(() => fetchMock.restore());
    
    test('should render the responsive container without children', async () => {
        fetchMock.post({ url: 'http://dummy.url' }, pieChartResponse);

        const { container } = render(<ResponsiveContainer
            setWidth={() => { }}
            height={300}
            api={{ url: 'http://dummy.url', params: {} }}
            setData={() => { }}
            fetchFnc={fetchFnc}
        />);

        await waitFor(() => container.querySelector('div'));

        expect(container).toMatchSnapshot()
    });

    test('should render the responsive container with children', async () => {
        fetchMock.post({ url: 'http://dummy.url' }, pieChartResponse);
        const { container, getByText } = render(<ResponsiveContainer
            setWidth={() => { }}
            height={300}
            api={{ url: 'http://dummy.url', params: {} }}
            setData={() => { }}
            fetchFnc={fetchFnc}
        >
            <p>This is a test children</p>
        </ResponsiveContainer>);

        await waitFor(() => container.querySelector('div'));
        expect(getByText("This is a test children")).toBeTruthy();
    });

    test('should set the data from the fetch function', async () => {
        fetchMock.post({ url: 'http://dummy.url' }, pieChartResponse);
        const setData = jest.fn();
        await act(async () => {
            render(<ResponsiveContainer
                setWidth={() => { }}
                height={300}
                api={{ url: 'http://dummy.url', params: {} }}
                setData={setData}
                fetchFnc={fetchFnc}
            >
                <p>This is a test children</p>
            </ResponsiveContainer>
            );
        });

        expect(setData.mock.calls.length).toBe(1);
    });

    test('should return error state', async () => {
        fetchMock.post({ url: 'http://dummy.url' }, 404);
        const { container, getByText } = render(<ResponsiveContainer
            setWidth={() => { }}
            height={300}
            api={{ url: 'http://dummy.url', params: {} }}
            setData={() => { }}
            fetchFnc={fetchFnc}
        >
            <p>This is a test children</p>
        </ResponsiveContainer>);

        await waitFor(() => container.querySelector('div'));
        expect(getByText("Error")).toBeTruthy();
    });
});