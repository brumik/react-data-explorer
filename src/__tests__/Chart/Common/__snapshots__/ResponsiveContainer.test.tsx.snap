// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart/Common/ResponsiveContainer should render the responsive container without children 1`] = `
<div>
  <div>
    <div
      style="height: 300px;"
    />
  </div>
</div>
`;
